### PR TITLE
Fixed binding element types coming from the out of tuple bounds under `noUncheckedIndexedAccess`

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/52302

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this

**Comment by Andarist** [comment]
@RyanCavanaugh do you happen to know if [this one](https://github.com/microsoft/TypeScript/pull/52318#issuecomment-1397822616) is just a flake or is it something that I should look into?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22767,7 +22767,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
         if (everyType(type, isTupleType)) {
-            return mapType(type, t => getRestTypeOfTupleType(t as TupleTypeReference) || undefinedType);
+            return mapType(type, t => {
+                const tupleType = t as TupleTypeReference;
+                const restType = getRestTypeOfTupleType(tupleType);
+                if (!restType) {
+                    return undefinedType;
+                }
+                const possibleOutOfBounds = index >= tupleType.target.fixedLength + getEndElementCount(tupleType.target, ElementFlags.Fixed);
+                return compilerOptions.noUncheckedIndexedAccess && possibleOutOfBounds ? getUnionType([restType, undefinedType]) : restType;
```
Comment (sandersn) [inline src/compiler/checker.ts:22777]: `getEndElementCount` iterates over each element of a tuple. I'd rather check noUncheckedIndexAccess first:

```suggestion
                if (compilerOptions.noUncheckedIndexedAccess &&                
                    index >= tupleType.target.fixedLength + getEndElementCount(tupleType.target, ElementFlags.Fixed) {                    
                    return getUnionType([restType, undefinedType]);                    
                }                
                else {                
                    return restType;                    
                }
```

Losing the name for `possibleOutOfBounds` isn't that bad; I think just about every programmer will recognise `index => ...` as checking for that.

**Review by sandersn — CHANGES_REQUESTED** [review]
Looks good, but can you reorder the oobounds calculation as requested?

**Review by sandersn — APPROVED** [review]

**Pull request merged**